pkg/action/hooks.go

```diff
@@ -83,6 +86,20 @@ func (cfg *Configuration) execHook(rl *release.Release, hook release.HookEvent,
 		err = cfg.KubeClient.WatchUntilReady(resources, timeout)
 		// Note the time of success/failure
 		h.LastRun.CompletedAt = helmtime.Now()
+
+		if isTestHook(h) {
+			// "Why create a new client for each hook?" Because `cfg` can be `nil` outside the `for` loop.
+			client, err := cfg.KubernetesClientSet()
```

Comment:
can you provide a test case for this? I have no context how this issue can occur. A link to existing kubernetes client documentation or an issue in client-go describing this behaviour would also suffice.

Also can't this be an issue where a not-insignificant amount of test hooks create a large amount of kubernetes clients? That seems... Sub-optimal. I'm sure there must be a better way to handle this.

Have you looked for existing kubernetes client implementations of the log endpoint for better ideas?

Reply:
For the record, the tests already fail if `client` is defined once outside the `for` loop.

<details>

```diff
diff --git a/pkg/action/hooks.go b/pkg/action/hooks.go
index 3d864e65..3d0a0871 100644
--- a/pkg/action/hooks.go
+++ b/pkg/action/hooks.go
@@ -44,6 +44,11 @@ func (cfg *Configuration) execHook(rl *release.Release, hook release.HookEvent,
        // hooke are pre-ordered by kind, so keep order stable
        sort.Stable(hookByWeight(executingHooks))
 
+       client, err := cfg.KubernetesClientSet()
+       if err != nil {
+               return errors.Wrapf(err, "unable to create Kubernetes client set to fetch pod logs")
+       }
+
        for _, h := range executingHooks {
                // Set default delete policy to before-hook-creation
                if h.DeletePolicies == nil || len(h.DeletePolicies) == 0 {
@@ -88,11 +93,6 @@ func (cfg *Configuration) execHook(rl *release.Release, hook release.HookEvent,
                h.LastRun.CompletedAt = helmtime.Now()
 
                if isTestHook(h) {
-                       // "Why create a new client for each hook?" Because `cfg` can be `nil` outside the `for` loop.
-                       client, err := cfg.KubernetesClientSet()
-                       if err != nil {
-                               return errors.Wrapf(err, "unable to create Kubernetes client set to fetch pod logs")
-                       }
                        hookLog, err := getHookLog(client, rl, h)
                        if err != nil {
                                return err
```

```
=== RUN   TestInstall/basic_install
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x20 pc=0x2e6db5f]

goroutine 461 [running]:
helm.sh/helm/v3/pkg/action.(*Configuration).KubernetesClientSet(0xc0006e7d40)
	/…/pkg/action/action.go:281 +0x3f
helm.sh/helm/v3/pkg/action.(*Configuration).execHook(0xc0006e7d40, 0xc00039ed90, {0x362c20a, 0x1}, 0xc000da2000)
	/…/pkg/action/hooks.go:47 +0x1a7
helm.sh/helm/v3/pkg/action.(*Install).performInstall(0xc000d1e9a0, 0xe05785, 0xc00039ed90, {0x0, 0x3a02b40, 0x0}, {0x565e208, 0xc00135ad40, 0x0})
	/…/pkg/action/install.go:358 +0xff
created by helm.sh/helm/v3/pkg/action.(*Install).RunWithContext
	/…/pkg/action/install.go:347 +0x1fa8
FAIL	helm.sh/helm/v3/cmd/helm	26.000s
```

```
=== RUN   TestInstallRelease
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x20 pc=0x2ea855f]

goroutine 37 [running]:
helm.sh/helm/v3/pkg/action.(*Configuration).KubernetesClientSet(0xc000340340)
	/…/pkg/action/action.go:281 +0x3f
helm.sh/helm/v3/pkg/action.(*Configuration).execHook(0xc000340340, 0xc000036f50, {0x33694ab, 0x100000000ebaf7b}, 0x0)
	/…/pkg/action/hooks.go:47 +0x1a7
helm.sh/helm/v3/pkg/action.(*Install).performInstall(0xc000350420, 0xc0000601f0, 0xc000036f50, {0x0, 0x0, 0x0}, {0x5238368, 0xc000060238, 0x0})
	/…/pkg/action/install.go:358 +0xff
created by helm.sh/helm/v3/pkg/action.(*Install).RunWithContext
	/…/pkg/action/install.go:347 +0x1fa8
FAIL	helm.sh/helm/v3/pkg/action	0.181s
```

</details>

Reply:
However, seems like, contrary what our comment claims, it is not `cfg` that is `nil` – it is in fact e.g. `&{<nil> 0xc000475240 0xc0004b4820 0xc000180870 0x47eb1a0 0x2eccec0}`.

Reply:
Currently investigating why the `TestInstallRelease` test fails with the above change (calling `cfg.KubernetesClientSet` before the `for` loop). Turns out `i.cfg.RESTClientGetter` is `nil` [when `i.cfg.execHook` is called in `performInstall` in `pkg/action/install.go`](https://github.com/helm/helm/blob/5d47255a3d6c886850d6c6cbead1b00b204d07b1/pkg/action/install.go#L358).

Reply:
> Also can't this be an issue where a not-insignificant amount of test hooks create a large amount of kubernetes clients? That seems... Sub-optimal.

On the other hand, a client is only ever needed when hook logs are to be acquired, i.e. only for test hooks. So while creating the client outside the `for` loop means we only create one, no matter the number of test hooks, it also means we _always_ create one, even when it's not needed at all.